### DISCUSSION TITLE
tutorials/lift-off-part3/02-updating-our-schema

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#18] [course feedback]
# tutorials/lift-off-part3/02-updating-our-schema

https://www.apollographql.com/tutorials/lift-off-part3/02-updating-our-schema

**Comment by manavm1990** [comment]
At 1:28, she says that the entries in the array can't be null either. But, if it's a new `Track`, won't it just be an empty array? I understand that the array itself is required, but I felt that it should be: `[Module]!`, no?

**Reply by manavm1990** [reply]
NM. I think I get it. We can't get `null` back. Instead, it will be an empty array and empty object as needed.

**Reply by mabuyo** [reply]
For clarification, these are valid return values for the `modules` field:

- `[]`
- `[{id: 'mod-1', title: 'Module 1: a title'}]`

And these are _invalid_ return values for the `modules` field:

- `null`
- `[null]`
- `[{id: 'mod-1', title: 'Module 1: a title'} , null, {id: 'mod-2', title: 'Module 2: another title'} ]`

To your point about the empty object, we defined in the schema that a `Module` type has the fields `id: ID!` and `title: String!` which are both _non-nullable_, so those two properties _do_ need to exist.

**Comment by jasonav** [comment]
Why did you not update TrackModel to add the description and numberOfViews properties?

**Reply by mabuyo** [reply]
Hey there @jasonav - great catch, we've just updated the course and starter repos to include this.

**Reply by neogeek13** [reply]
I am assuming it was this late addition but this is the first time I've seen a video not cover all the content in the text below - specifically the Updating our models section updating the models and codegen files. Hopefully that will be coming soon.

P.S. @mabuyo I love your Hello World t-shirt. Dunno if it's available online somewhere but if so would love a link. <3

**Reply by mabuyo** [reply]
@neogeek13 The TypeScript content is fairly recent and it's on our list to produce videos for it!

Thank you, I love this shirt! [I got it from bubblesort.io](https://shop.bubblesort.io/products/hello-world-tshirt?variant=7123289800740) - unfortunately they're all sold out. The creator does limited merch drops and I'm always on the lookout for them (I have several of her other shirts & jackets! 👀 )

**Comment by Chip-L** [comment]
Maybe we will cover this later and I haven't gotten to the fragments part yet. But, when we require a non-null modules array, I get this error in `.\client\src\pages\tracks.tsx`:

>  Property 'modules' is missing in type '{ __typename?: "Track" | undefined; id: string; title: string; thumbnail?: string | null | undefined; length?: number | null | undefined; modulesCount?: number | null | undefined; author: { ...; }; }' but required in type 'Track'.

As soon as I take make the modules array optional (`modules: [Module!]`) in my schema, it fixes the error. I also noticed in the `Track` type in `grapql.ts`, the modules property goes from being required (`modules: Array<Module>`) to not required (`modules?: Maybe<Array<Module>>`).

**Reply by mabuyo** [reply]
Hi @Chip-L , are you working from your Lift-off II code repo or did you clone the `lift-off-part3` repo? If you worked off last course's repo, could you try running `npm run generate`?

**Reply by Chip-L** [reply]
@mabuyo, I actually started in Lift-off I and worked up through II and III. I have ran `generate` on both the server and client. I've watched how things change in the `.\client\src\__generated__\graphql.ts` when I change it from required to not-required. 

I actually think it is correct that we get an error there. The type being returned, per TypeScript,  expects to ALWAYS have the array. We don't return the array as part of the `GetTracks` query, but our `Track` type is still expecting it:
```
// graphql.ts

export type Track = {
  __typename?: 'Track';
  author: Author;
  description?: Maybe<Scalars['String']['output']>;
  id: Scalars['ID']['output'];
  length?: Maybe<Scalars['Int']['output']>;
  modules: Array<Module>;  //<--------------- Field is expected
  modulesCount?: Maybe<Scalars['Int']['output']>;
  numberOfViews?: Maybe<Scalars['Int']['output']>;
  thumbnail?: Maybe<Scalars['String']['output']>;
  title: Scalars['String']['output'];
};
```

The only ways around this that I can see is:
 - don't make the module array required in the `typeDefs`
 - or create a second type

I don't see that there is a way to extend our GraphQL types or do unions.

**Comment by naskonaskov** [comment]
In previous parts of the lift-off we created the mapper in the codegen including the track map, because the object we get from the rest api was different from the one in our schema (authorId and list of authors). Later we added mapper configuration for Author and now we are adding for Module, but the object returned from the rest api has the same properties we have defined in our schema. Why is it needed?
